### DISCUSSION TITLE
Will codon support existing Python3 tooling?

### DISCUSSION THREAD

**Discussion by phrmoy** [exaloop/codon#370] [General]
For adoption, it's important that the tremendous amount of effort poured into the existing Python3 ecosystem be leveraged to the fullest. For example, lately, I started incorporating Cython into my workflow because in "Pure Python Mode," I can write Python, and use all my existing tooling (Pylance, Black, mypy, etc). Writing actual Cython code was never bad necessarily but foregoing the wealth of VSCode support made me look at alternatives such as Numba (which is essentially Python code and thus leveraged all of my tools to the fullest and seamlessly).

Let me give out another example, I think Crayon's new HPC language Chapel is badass. However, IMHO, I believe that no matter how badass the language is, programming is a group activity. Now while research types may be willing to take more risks than business types, whether I am doing research or business (I am associated with both), I will pick whichever close alternative gets me what I need with the most effort and maintenance being put into it (this necessarily includes both the compiler and all the surrounding tooling for development -- libs may or may not be a show stopper). In this particular Chapel case, I ended up picking Julia. At the time, I wanted to go with Chapel (for personal preferences), but Julia seemed like the more appropriate choice given how far along the tooling was (libs too).

Now while it's understandable that libraries need to flourish over time (if you can hack into Python libs, maybe you can? Bonus points), it seems you guys have a strong case to pick up on much if not all of the existing tooling surrounding Python3.

Rooting for your success!

Pedro

**Comment by elisbyberi** [comment]
Hi @phrmoy,

I use the same environment to develop for Codon as I do for CPython. This means that I test the code in CPython and then compile it with Codon. By the way, could you please clarify which specific tools you are referring to?

**Reply by phrmoy** [reply]
Pylance, Black, mypy, sonarlint, and some other linting tools. If you're already using these tools without issues in the development of Codon code, that is super awesome! Would that also mean that you get to run you codon code as a non-compiled version in CPython?

**Reply by elisbyberi** [reply]
> Would that also mean that you get to run you codon code as a non-compiled version in CPython?

That's correct. I test the code in CPython first, and then I compile it using Codon by simply adding the `@codon.jit` decorator to the function I want to compile.

By using a  [Python decorator](https://docs.exaloop.io/codon/interoperability/decorator), you can write code that is 100% compatible with CPython. However, only functions decorated with `@codon.jit` will be compiled by Codon. In my opinion, that is the best way to make use of the Codon compiler.

Additionally, you can create Python C extensions using Codon.

Codon offers some advanced programming features that are not supported by the CPython tooling. It would be beneficial to address this issue.
